### PR TITLE
Chunkable token classification pipeline

### PR DESCRIPTION
**Pull request by luccailliau**
This PR improve the TokenClassificationPipeline by extending its usage to tokenized texts longer than `model_max_length` by returning overflowing tokens as chunks rather than truncating texts. To enable the use of this extended feature, you must use a fast tokenizer with an aggregation strategy different to `"none"` and provide a `stride` number.

# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by LucCailliau** [comment]
For whoever is reading this, I will quickly correct the alignment issues which concerns tokens that are removed due to special tokens mask :)

**Comment by sgugger** [comment]
cc @Narsil
I'll let you review and decide if it maybe would be more suitable as code on the hub or needs to be in Transformers.

**Comment by Narsil** [comment]
Thank you for this PR. It looks promising.

> is now able to process sequences longer than 512.

Do you have a specific model in mind, `512` seems oddly specific.

> if it maybe would be more suitable as code on the hub or needs to be in Transformers.

I think it really depends on the complexity of the resulting code, and the cross over with other parameters.

@LucCailliau the tests aren't passing yet. I don't want to do a full review before the tests are passing.


Some notes:
- Existing tests cannot be modified and must be passing
- Overall code should be relatively simple (the current state looks ok on the surface).
- The most complex part (I think it should be the conflict resolution in overlapping parts) must be as clear as possible. It's ok to put it into another function for special handling.
- Unless it causes an explosion in complexity, it should work with all `aggregation_strategy`. The minimum are `NONE` and `SIMPLE`. If it causes and explosion in complexity, we need to forbid the use of the unsupported combinations
- We need good tests for this feature:
  -  Make sure it actually solves what this PR is set to do (handler longer than `model_max_length` inputs`). (Both a slow and fast test)
  - Check with `aggregation_strategy` parameters.
  - Check errors if preventing some parameter combo.
  
Does this make sense ?

For the sake of getting this moving forward faster I actually suggest splitting this over several PRs.

The first PR should be the move to `ChunkPipeline` and nothing else should be modified. Then adding the new parameter. We would need the second one, to be close enough to good state to merge the first (there's no point in changing the pipeline type if we're not able to support this `process_all` parameter correctly.

Don't hesitate to ask more question should you need it.

**Comment by LucCailliau** [comment]
@Narsil, all tests passed except the code quality. I used black but it doesn't pass. I also update the schema above to explain the algorithm for update/aggregate scores

**Comment by Narsil** [comment]
> @Narsil, all tests passed except the code quality. I used black but it doesn't pass. I also update the schema above to explain the algorithm for update/aggregate scores

try  `pip install -e .[quality] && make fixup` to use the correct black version.

**Comment by LucCailliau** [comment]
@Narsil, it doesn't work

**Comment by LucCailliau** [comment]
@Narsil done

**Comment by LucCailliau** [comment]
Simple to use
`pipe = TokenClassificationPipeline(
    model=model,
    tokenizer=tokenizer,
    aggregation_strategy="simple",
    process_all=True,
    stride=64
)`

**Comment by LucCailliau** [comment]
@Narsil, wait, just one more thing to add and we can go

**Comment by LucCailliau** [comment]
@Narsil, I checked everything on my side, we can go for it :)

File: src/transformers/pipelines/token_classification.py
Comment (Narsil) [inline src/transformers/pipelines/token_classification.py:196]: Don't sanitize here.

negative stride is a bug, but it's not your error to handle, this is a tokenizer level bug.

File: src/transformers/pipelines/token_classification.py
Comment (Narsil) [inline src/transformers/pipelines/token_classification.py:201]:
```suggestion
                if stride is None:
                    stride = 0
```

File: src/transformers/pipelines/token_classification.py
Comment (Narsil) [inline src/transformers/pipelines/token_classification.py:205]: Remove `padding`.

File: src/transformers/pipelines/token_classification.py
Comment (Narsil) [inline src/transformers/pipelines/token_classification.py:204]: Remove this.

In combination with padding it's a HUGE performance loss, potentially crashing things.
Some tokenizer do not have a max_length and this is set to basically infinity, using this with padding has potential to blow memory up.

If the tokenizer handles infinite lengths, there's no need to chunk anything.

**Review by Narsil — COMMENTED** [review]
[CRITICAL] 

I think the logic for this is actually flawed. See relevant comment


We need to update the docs. And add **several** very readable tests where this feature is showcased.

We need at least 1 evidently enabling case (raised error before, now passes through and output looks legit.) This can be done by artificially reducing the model_max_length of a model.


Then we need several ways to showcase what happens for a *merge*. 

Let's consider the example. "I live in New York and I like it".
And have the stride be over "New York" entirely.

Then consider several cases. What happens if left hand side considers this isn't a loc and right hand side does.
What if both agree that "New" is loc and "York" is nothing.

And a few real large test cases with real models.

File: src/transformers/pipelines/token_classification.py
Comment (LucCailliau) [reply src/transformers/pipelines/token_classification.py:199]: If you remove padding here, you'll have an error: `ValueError: Unable to create tensor, you should probably activate truncation and/or padding with 'padding=True' 'truncation=True' to have batched tensors with the same length. Perhaps your features (`input_ids` in this case) have excessive nesting (inputs type `list` where type `int` is expected).` That's why I kept it

File: src/transformers/pipelines/token_classification.py
Comment (LucCailliau) [reply src/transformers/pipelines/token_classification.py:327]: "Here you are mix and matching hopeful that a more "confident" a token is ,the more likely it is but that's simply not true, adversarial attacks show that plenty." => Of course, but we have to choose one or maybe get entities independently for each chunk and merge results after this step. Again we'll have to choose between entities with different scores.

In fact even with the current implementation, if you provide a text with New York Bagels at the end with Bagels truncated, the model will be wrong due to the lack of context:
image


There is no perfect method of course. Keep the highest score is an acceptable solution I think or maybe keep the highest scores for the aggregated entities at the final step. I don't think we can do a lot more for that, we are restricted with the model's architecture. 

Maybe we can create a validation dataset to validate the approach? It's a nice feature for companies dealing with long texts.

File: src/transformers/pipelines/token_classification.py
Comment (Narsil) [reply src/transformers/pipelines/token_classification.py:207]: Ok, we may need `padding=True`. I thought it wasn't but it could. The main thing to look out for is getting huge batch filled only with padding tokens.

File: src/transformers/pipelines/token_classification.py
Comment (Narsil) [reply src/transformers/pipelines/token_classification.py:279]: No the previous code was selectively popping `offset_mapping` and `sentence`. Which is not the same as this, which relies on `self.tokenizer.model_input_names` to be correct. 

The issue with that approach, is that should `model_input_names` fail to mention `attention_mask` for instance, then it would silently start to behave incorrectly (by failing to mask the proper input_ids).

`model.forward(**tokenizer(string))`  is the invariant we should upkeep.

Since we have to add more than necessary for the post processing, we are in control of which to remove too in the pipeline.
That way we don't have on tokenizer internals (which could and unfortunatly sometimes are incorrect).

Do you see the difference now ?

File: src/transformers/pipelines/token_classification.py
Comment (Narsil) [reply src/transformers/pipelines/token_classification.py:294]: If tokenizer doesn't have `max_length` then the `stride` argument shouldn't be used because the model doesn't require it, it can do without chunking.

`stride=None` is perfectly correct for this default here.

For the unlimited lengthe tokenizer, we should raise an error instead. (Maybe chunking could still be desired for performance reasons in which case I think users should artificially limit the tokenizer for this to work, wdyt ?)

File: src/transformers/pipelines/token_classification.py
Comment (LucCailliau) [reply src/transformers/pipelines/token_classification.py:279]: Yes I understand, here we rely only on the tokenizer. I'll revert back to the previous function. (This function dosn't need to be modified)

File: src/transformers/pipelines/token_classification.py
Comment (Narsil) [reply src/transformers/pipelines/token_classification.py:335]: > There is no perfect method of course.

Entirely correct, which is why it cannot be on by default (we can't hide those tricks for users).

> I don't think we can do a lot more for that, 


I just proposed one above

> A very simple merge logic, could be to simply split the stride in half, use the first logits for the left part, and use the second logits for the next one. But this is still keeping the issue in the middle of the stride, but we can hope that in that case since both parts see the 2 tokens in a majority of time both chunks will agree. That needs to be heavily checked though.

I am not sure which is better, which is why we need some benchmarks real testing to gauge pros&cons. 
Like testing a few different (ideally different architectures) on a few different datasets. 
If you don't have the ressources to run complete benchmarks, you could stop at creating the scripts with extremely tiny amount of data and we could run it on larger data. 

I think here we need to be very pragmatic. I have concerns about your proposal from my theoretical understanding of models, but if it gets better results on representative tests, I'd be more than happy to be proven wrong.

My proposed approach seems naïve but is also relatively simple which is a huge strength in your proposal. 
I think aiming for simple and orthogonal to aggregation_strategy is really nice !

Comment (LucCailliau) [reply src/transformers/pipelines/token_classification.py:335]: I'm also experimenting a new approach:
- pass each chunk to aggregate function (without striding them) and get entities
- filter entities on the start index and keep the longest
I'll let you know

Comment (Narsil) [reply src/transformers/pipelines/token_classification.py]: Yes. I think any solution to this problem really requires the test I'm mentioning.

It's a bit too hard in my experience to anticipate how a strategy will preform.
Several architecture avoids overfitting a particular model, several datasets avoids overfitting English and a particular dataset.

I don't see any way around it.

**Comment by Narsil** [comment]
> I'll start creating a validation set to have results of what we've done.

Sounds great ! Again don't hesitate to ask for resources for larger runs.

**Comment by LucCailliau** [comment]
> > I'll start creating a validation set to have results of what we've done.
> 
> Sounds great ! Again don't hesitate to ask for resources for larger runs.

Great, do you have a specific model and/or dataset to perform our tests? I am also interested for resources.

The tests must be done with aggregation strategy to "simple", "first" and "average". If "none" is set we can have the following:
"We went to Manhattan Burgers, it was awesome!"
- "Manhattan" -> "B-LOC"
- "Burgers" -> "I-ORG"
And here again we can update scores as we did previously, but this is not perfect.
Corrections can only be applied if `aggregation_strategy` different from `None`.

**Comment by LucCailliau** [comment]
@Narsil, I finished the tests for this pipeline and the results are convincing since it improves the current implementation.

### A summary of the PR:

This PR improve the TokenClassificationPipeline by extending its usage to tokenized texts longer than `model_max_length` by return overflowing tokens as chunks rather than truncating texts. To enable the use of this extended feature, you must use a fast tokenizer with an aggregation strategy different to `"none"`.

### Approachs:

Different approaches have been experimented: 

1. Updating tokens scores in overlapping parts.
2. Aggregate entities for all chunks, no matter the overlapping parts.

The first approach which consist of updating scores with the highest is not the best. A more "confident" a token doesn't mean the more likely it is, adversarial attacks show that plenty.

The second approach (selected approach) consist of processing each chunk and aggregate entities no matter overlapping part. In the final aggregation step, we select the best entity in overlapping parts with a rule. We first look at the longest entity and if entities have the same length, we take the entity with the highest score.

Example:
Given the following entities from their respective chunk:
"New York" -> "LOC": from chunk no. 1
"New York City" -> "LOC": from chunk no. 2
The remaining entity in aggregated entities will be "New York City" -> "LOC"

### Results

In order to compare the current implementation with this one, we generated labeled text from the conll2003 dataset (available on the hub). Then, compared the number of exact match and wrong match only in the first chunk for each implementation.

For `aggregation_strategy` set to `simple` we have in 415 texts:

- **12856** exact matchs and **2111** wrong matchs for the proposed implementation
- **12739** exact matchs and **2181** wrong matchs for the current implementation

As chunked method is able to correct entities with more context, we increased the number of exact matchs by 117 and reduced the number of wrong match by 70.

### Next steps

I'm finalizing the full study. It will be available soon in this section as a notebook explaining each steps.
@Narsil, can you also check with the check_code_quality since the error comes from another file which is not related to this commit?

**Review by Narsil — COMMENTED** [review]
Overall very good PR given the value it provides.

Lots of nits everywhere.

We need to add tests to confirm and make sure this approach is kept intact.

I suggest using my following script in tests to make a small model have a super small range, and give it sentences to showcase the capacity to infer on infinite range + the aggregation strategy.
And the same (could be smaller test) with a large test with an interesting example (like "New York" happening in the stride for instance, bonus points if both sides of the stride don't put the same value on them)

For the testing part, I tried your script however I feel like it's not doing a fair job.
The problem I see is that you're comparing the model which cannot do the job, to your extension.

What would really check the value of your proposal IMO is to run on sentences short enough to be run by the model without chunking+striding AND then force chunking into smaller parts, and see how your similar your pipeline performs on it.

Here is the script I propose:

```python
from datasets import load_dataset

from transformers import pipeline


model_id = "StanfordAIMI/stanford-deidentifier-base"
# model_id = "Davlan/distilbert-base-multilingual-cased-ner-hrl"

MAX_LENGTH = 50
STRIDE = 10


pipe = pipeline(model=model_id, aggregation_strategy="first", device="cuda:0", batch_size=50)

chunked_pipe = pipeline(model=model_id, aggregation_strategy="first", stride=STRIDE, device="cuda:0", batch_size=50)
chunked_pipe.tokenizer.model_max_length = MAX_LENGTH

dataset_names = [("conll2003",), ("wikiann", "fr")]
for dataset_name in dataset_names:
    dataset = load_dataset(*dataset_name)

    def text_(example):
        example["text"] = " ".join(example["tokens"])
        return example

    ds = dataset["train"]
    ds = ds.map(text_)

    def stream():
        for example in ds:
            yield example["text"]

    chunked_pipe(ds[0]["text"])

    total = 0
    correct = 0
    uninteresting = 0
    for i, (regular_output, chunked_output) in enumerate(zip(pipe(stream()), chunked_pipe(stream()))):
        if regular_output == chunked_output:
            uninteresting += 1
        else:
            for item in regular_output:
                item.pop("score")
            for item in chunked_output:
                item.pop("score")
            if regular_output == chunked_output:
                correct += 1
            else:
                pass
                # for item, chunked_item in zip(regular_output, chunked_output):
                #     if item != chunked_item:
                #         print(item, "!=", chunked_item)
            total += 1

    print(f"Results : {correct/total:.2f} ({correct}/{total} (ignored {uninteresting})")
```

Several notes:
- The smaller the MAX_LENGTH, the worse it performs (naturally).
- batching done in the pipeline (to benefit from GPU acceleration)
- Using strategy `first` rather than `none`.  Since I'm running on models on unseen data, they seem to perform really bad naturally (B- , B- Instead of B-, I- on a lot of tokens). Using `first` allows a first pass of cleanup for both normal and chunked pipelines to limit this effect in the testing.
- Not skipping too long strings in the dataset directly, because the tokenizer can do cleverer things because of the [BOS, EOS] tokens that can get added automatically with striding, making it a hard strategy.
- Ignored values, probably means there was no NER within the stride, making the resolution strategy uninteresting.
- There's an optional printing to see the different tokens. I looked at it , and didn't see anything shocking, some spans are a bit different, some group are different and without context it's hard to tell who's correct.

Here is the result on this exact script:

```python
Results : 0.95 (1461/1540 (ignored 1913)
Results : 0.98 (3018/3070 (ignored 6930)
```

Do you agree with the proposed test for actually checking how it performs on stride conflicts ?

I'm now going to expand the tests a bit on a bit more dataset (more languages) and more models.

I'll wait on the nits+tests to be fixed before calling a core maintainer.

**Comment by Narsil** [comment]
And here are more complete logs so you can inspect a bit more the edgy cases if you want:

wikiann all languages x   top 5 token classification models

Overall it seems good enough to add to me.

**Comment by luccailliau** [comment]
> Do you agree with the proposed test for actually checking how it performs on stride conflicts ?

It's a good approach too. It makes more sense since concatenate sentences can lead to new randomly generated sentences.
However, with this new script, we need to be careful of:

- most of sentences in datasets fit in a sequence of length 50 (`uninteresting` is incremented when it's the case)
- we take as references results of pipeline without chunking+striding and not the references from the dataset itself
- only one non-detected entity in a sentence with the pipeline chunking+striding will cause an entity offset and lead to wrong results

**Comment by luccailliau** [comment]
> * There's an optional printing to see the different tokens. I looked at it , and didn't see anything shocking, some spans are a bit different, some group are different and without context it's hard to tell who's correct.

I noticed the same behavior with the previous tests. In our case, the aim is not to apply corrections on predictions but to aggregate entities. If the model is 100% correct, the aggregated entities will also be 100% correct.

**Comment by luccailliau** [comment]
> And here are more complete logs so you can inspect a bit more the edgy cases if you want:

I think you forgot to link the results

> wikiann all languages x top 5 token classification models
> 
> Overall it seems good enough to add to me.

Yes, with the previous tests and with manual tests on different content, it's working as expected.

**Comment by luccailliau** [comment]
I refactored the code as you mentioned. All is good, on my side

**Comment by Narsil** [comment]
Oops: https://gist.github.com/Narsil/e8609805e8e52c7e4114586eede8a481

**Comment by luccailliau** [comment]
> Oops: https://gist.github.com/Narsil/e8609805e8e52c7e4114586eede8a481

Good results!

File: src/transformers/pipelines/token_classification.py
Comment (Narsil) [reply src/transformers/pipelines/token_classification.py:330]: No no you can apply it all the time, if there is only 1 chunk the function should naturally "ignore" everything.

Like there cannot be an "overlap" that you're checking for in `aggregate_entities` if there's only 1 chunk.

The benefit to remove this is that it makes logic easier to understand.

**Review by Narsil — COMMENTED** [review]
Thanks for all the modifications.

Do you want help creating the tests ?

@sgugger for semi final review. (Still need to add some tests before merging).

File: src/transformers/pipelines/token_classification.py
Comment (luccailliau) [reply src/transformers/pipelines/token_classification.py:337]: @Narsil, here is an example if you remove `sorted()` and you have entities in the overlapping part.
I took the Wikipedia article of Elon Musk (we have lot of entities to visualize) and comment this line:
`# entities = sorted(entities, key=lambda x: x["start"])`
The aggregation function is not able to work anymore now. We have duplicates entities. Like SolarCity, it is present twice:
image

File: src/transformers/pipelines/token_classification.py
Comment (luccailliau) [reply src/transformers/pipelines/token_classification.py:330]: Ok, I'll test it and see if it pass the tests

File: src/transformers/pipelines/token_classification.py
Comment (sgugger) [inline src/transformers/pipelines/token_classification.py:268]:
```suggestion
            if offset_mapping is not None:
```
This should have been in the code before. This will throw an error if `offset_mapping` is a tensor or a numpy array.

**Review by sgugger — COMMENTED** [review]
Thanks for working on this. I'm in line with @Narsil comments on this.

File: src/transformers/pipelines/token_classification.py
Comment (luccailliau) [reply src/transformers/pipelines/token_classification.py:337]: @Narsil, for performance reasons regarding `aggregate_overlapping_entities()`, I don't think we can do better in pure python. I understand that at first glance this function seems to be not optimized due to the `sorted()` method but in fact it is.

The number of entities in overlapping parts can be different between chunks. We can't be sure we don't have an overlapping entity until we look at the next one. Entities must be sorted for this.

For some specific applications where you need the full capability of the transformers with the widest context, you can choose a stride number greater than `model_max_length / 2` and in that case you will have the same entity in many chunks and it is important to tackle this case.

Even if you don't want to sort the entities, you don't know where the overlapping entities are and that will end up with different loops (at least O(n²) complexity) and method calls, which is time consuming. Whereas now with `sorted()` + the for loop we have an average time complexity of O(2n), because `sorted()` is already deeply optimized.

As you can see below, even for 100000 entities retrieved in one text, the function only takes 40ms to aggregate them. The calculation time of all these chunks to extract these 100000 entities will be much more important than these 40ms.
image

**Comment by luccailliau** [comment]
@Narsil, I updated the comments in the Files Changed section that give a better explanation of the `aggregate_overlapping_entities()` method. I don't know if you received a notification for that. Is it good for you?

**Comment by Narsil** [comment]
I didn't but we're still missing some tests though.

I offered to help writing them if you want, but we really want tests to show case this feature (and make sure it doesn't break later).

The most important will be small tests (with tiny models) so that they run always.
And showcase what's happening on simple strings while setting `model_max_length` to something super tiny to force the chunking to happen.

**Comment by luccailliau** [comment]
I thought it was for @sgugger since you already created a script that shows it works. @Narsil your help is welcome. I have in mind to create specific tests with manually checked references to ensure it doesn't break later.

**Comment by Narsil** [comment]
No I pinged him because the code was clean enough to be looked at, but the tests (especially for such a complex feature) are mandatory.

If you can get started that would be immensely appreciated, but please share if you're having a hard time or don't have the bandwidth to do it. It shouldn't take me that much time but it's always better if you can write the tests yourself.

**Comment by luccailliau** [comment]
Great, I'll do it

**Comment by luccailliau** [comment]
@Narsil, I finished the tests. Ready for final review

**Comment by luccailliau** [comment]
@Narsil, I finished the tests.
The selected model is `elastic/distilbert-base-uncased-finetuned-conll03-english` (pytorch_model.bin of size 266MB). It is not a tiny model as you recommend. I tried different tiny models but unfortunately, they don't perform well on our example. In fact, it is not noticeable when running since it requires between 2s and 3s to run all the new tests.

The tests are composed in two parts:

1. `test_chunking()`: Test the pipeline with all aggregation strategies and match an hard coding output
2. `test_regular_chunk()`: Test the pipeline with all aggregation strategies and match output (without scores) between regular output (without chunking+striding) and chunked output (with chunking+striding)

The second test, `test_regular_chunk()` can be optional since the first, `test_chunking()` was created with the same rule: regular output match chunked output. But, even if the regular output should not be interpreted as a reference, in this case, it is good to show that we didn't create a flawed test.

The selected parameters for these tests are:
- `model_max_length=10`: extremely tiny to increase the difficulty
- `stride=5`: quite large (regarding `model_max_length`) to generate multiple chunks (15 chunks in our example)

You can also find below the output without `aggregate_overlapping_entities()` (but sorted by `"start"` for more readability) to see how the tests cover different overlapping cases:
 
image

And below, the same output with `aggregate_overlapping_entities()`:

image

Is it good for you?

**Comment by luccailliau** [comment]
Hello @Narsil, I don't know if you didn't receive the update for the tests or if you don't have time to look at it. Do not hesitate if you need additional work to be done

**Review by Narsil — APPROVED** [review]
LGTM.

@luccailliau 
I took the liberty of adding directly the small test I had in mind.
and pushing the other 2 tests to slow tests (since they use real models).
They are good tests btw !

@sgugger If you're ok for final review.

File: src/transformers/pipelines/token_classification.py
Comment (sgugger) [inline src/transformers/pipelines/token_classification.py:69]:
```suggestion
        stride (`int`, *optional*):
```
We don't write the default if it's None as a convention in the repo.

File: src/transformers/pipelines/token_classification.py
Comment (sgugger) [inline src/transformers/pipelines/token_classification.py:213]:
```suggestion
                    raise ValueError(
                        "`stride` was provided to process all the text but you're using a slow tokenizer."
                        " Please use a fast tokenizer."
```

File: src/transformers/pipelines/token_classification.py
Comment (sgugger) [inline src/transformers/pipelines/token_classification.py:420]: Please put this back on one line.

**Review by sgugger — COMMENTED** [review]
Thanks for all the work done on this! I just have a couple of last comments on styling and doc.

**Comment by luccailliau** [comment]
@Narsil, thank you very much for your help and your time on this PR, it was a pleasure!
@sgugger, I updated your changes, ready to go

**Comment by sgugger** [comment]
It looks like two of the comments have been resolved without any changes. Did you forget a commit maybe?

**Comment by luccailliau** [comment]
> It looks like two of the comments have been resolved without any changes. Did you forget a commit maybe?

Oops, should be ok now

**Comment by sgugger** [comment]
Perfect, congrats on finishing this big PR!

**Comment by luccailliau** [comment]
Thanks!

**Comment by tuanardouin** [comment]
Amazing work, thanks @luccailliau !

**Comment by aiancheruk** [comment]
Great work on this PR @luccailliau, congrats!

**Comment by Narsil** [comment]
Thanks for the great work !

**Comment by luccailliau** [comment]
Thank you all!

**Pull request merged**
